sdk/metric/periodic_reader.go

```diff
@@ -218,25 +232,61 @@ func (r *periodicReader) Collect(ctx context.Context) (metricdata.ResourceMetric
 	return ph.produce(ctx)
 }
 
-// ForceFlush flushes the Exporter.
+// export exports metric data m using r's exporter.
+func (r *periodicReader) export(ctx context.Context, m metricdata.ResourceMetrics) error {
+	c, cancel := context.WithTimeout(ctx, r.timeout)
+	defer cancel()
+	return r.exporter.Export(c, m)
+}
+
+// ForceFlush flushes pending telemetry.
 func (r *periodicReader) ForceFlush(ctx context.Context) error {
+	errCh := make(chan error, 1)
+	select {
+	case r.flushCh <- errCh:
+		select {
+		case err := <-errCh:
+			if err != nil {
+				return err
+			}
+			close(errCh)
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	case <-r.done:
+		return ErrReaderShutdown
+	case <-ctx.Done():
+		return ctx.Err()
+	}
 	return r.exporter.ForceFlush(ctx)
 }
 
-// Shutdown stops the export pipeline.
+// Shutdown flushes pending telemetry and then stops the export pipeline.
 func (r *periodicReader) Shutdown(ctx context.Context) error {
 	err := ErrReaderShutdown
 	r.shutdownOnce.Do(func() {
 		// Stop the run loop.
 		r.cancel()
-		r.wg.Wait()
+		<-r.done
 
 		// Any future call to Collect will now return ErrReaderShutdown.
-		r.producer.Store(produceHolder{
+		ph := r.producer.Swap(produceHolder{
 			produce: shutdownProducer{}.produce,
 		})
 
-		err = r.exporter.Shutdown(ctx)
+		if ph != nil { // Reader was registered.
```

Comment:
We care if `export` is called concurrently, but not `collect`.

`collect` has no shared state (it could be a stand alone function, but I wanted to scope its significance to the periodic reader so made it a method), and only calls `produce` of a producer. `produce` is [defined to be concurrent safe](https://github.com/open-telemetry/opentelemetry-go/blob/d7bfe6675fe79b0c01cae3d4c5f06330b2fece73/sdk/metric/reader.go#L91-L94).

There are two places where `export` is called: in `run` (via `collectAndExport`) and here (I made sure that `ForceFlush` delegates its access to these methods via `run` to avoid concurrent calls). At this point in `Shutdown` we are guaranteed that `run` has completed (based on `r.done` being closed) therefore the call to `export` will never be called concurrently.

For what it's worth, [our reader tests](https://github.com/open-telemetry/opentelemetry-go/blob/d7bfe6675fe79b0c01cae3d4c5f06330b2fece73/sdk/metric/reader_test.go#L110-L139) did catch this when I initially didn't delegate from `ForceFlush`.

Reply:
Ah, got it.  I see now that cancel() cancels run.